File: python_client/files/share_manager.py

from pathlib import Path


class ShareManager:

    def __init__(self, shared_dir: Path, shared_vault_store, share_index_store):
        self.shared_dir = shared_dir
        self.shared_vault_store = shared_vault_store
        self.share_index_store = share_index_store
        self.shared_dir.mkdir(parents=True, exist_ok=True)
        self._migrate_plaintext_shares()

    def list_files(self):
        index = self.share_index_store.load()
        return sorted(index.keys())

    def add_file(self, source_path: Path):
        filename = source_path.name
        self.shared_vault_store.store_file(filename, source_path.read_bytes())
        index = self.share_index_store.load()
        index[filename] = {"stored_in": "shared_vault"}
        self.share_index_store.save(index)
        plaintext_path = self.shared_dir / filename
        if plaintext_path.exists():
            plaintext_path.unlink()

    def has_file(self, filename: str):
        index = self.share_index_store.load()
        return filename in index

    def get_file_bytes(self, filename: str) -> bytes:
        if not self.has_file(filename):
            raise FileNotFoundError(filename)
        return self.shared_vault_store.get_file(filename)

    def get_file_size(self, filename: str) -> int:
        return len(self.get_file_bytes(filename))

    def _migrate_plaintext_shares(self):
        index = self.share_index_store.load()
        changed = False
        for path in self.shared_dir.iterdir():
            if not path.is_file():
                continue
            filename = path.name
            self.shared_vault_store.store_file(filename, path.read_bytes())
            index[filename] = {"stored_in": "shared_vault"}
            path.unlink()
            changed = True
        if changed:
            self.share_index_store.save(index)

File: java_client/src/main/java/com/cisc468share/files/ShareManager.java
package com.cisc468share.files;

import java.nio.file.Files;
import java.nio.file.Path;
import java.util.List;

import com.cisc468share.storage.ShareIndexStore;
import com.cisc468share.storage.VaultStore;

import java.util.LinkedHashMap;
import java.util.Map;

/**
 * Manages file sharing operations.
 */
public class ShareManager {

    private final Path sharedDir;
    private final VaultStore sharedVaultStore;
    private final ShareIndexStore shareIndexStore;

    public ShareManager(Path sharedDir, VaultStore sharedVaultStore, ShareIndexStore shareIndexStore) {
        this.sharedDir = sharedDir;
        this.sharedVaultStore = sharedVaultStore;
        this.shareIndexStore = shareIndexStore;
        try {
            Files.createDirectories(sharedDir);
            migratePlaintextShares();
        } catch (Exception e) {
            throw new RuntimeException("Failed to create shared directory", e);
        }
    }

    public List<String> listFiles() {
        return shareIndexStore.listShares();
    }

    public Path getFilePath(String filename) {
        return sharedDir.resolve(filename);
    }

    public boolean hasFile(String filename) {
        return shareIndexStore.getShare(filename) != null;
    }

    public void addFile(Path source) throws Exception {
        String filename = source.getFileName().toString();
        sharedVaultStore.storeFile(filename, Files.readAllBytes(source));
        Map<String, Object> info = new LinkedHashMap<>();
        info.put("stored_in", "shared_vault");
        shareIndexStore.addShare(filename, info);
        Files.deleteIfExists(sharedDir.resolve(filename));
    }

    public byte[] getFileBytes(String filename) {
        if (!hasFile(filename)) {
            throw new RuntimeException("Shared file not found: " + filename);
        }
        return sharedVaultStore.getFile(filename);
    }

    public long getFileSize(String filename) {
        return getFileBytes(filename).length;
    }

    public Path getSharedDir() {
        return sharedDir;
    }

    private void migratePlaintextShares() throws Exception {
        if (!Files.exists(sharedDir)) {
            return;
        }

        try (var paths = Files.list(sharedDir)) {
            paths.filter(Files::isRegularFile).forEach(path -> {
                try {
                    String filename = path.getFileName().toString();
                    sharedVaultStore.storeFile(filename, Files.readAllBytes(path));
                    Map<String, Object> info = new LinkedHashMap<>();
                    info.put("stored_in", "shared_vault");
                    shareIndexStore.addShare(filename, info);
                    Files.deleteIfExists(path);
                } catch (Exception e) {
                    throw new RuntimeException("Failed to migrate shared file", e);
                }
            });
        }
    }
}
